docs/dificil/small_to_large.md

```diff
@@ -1,7 +1,233 @@
+
+Cu această schimbare, codul va arăta așa:
+
+```c++
+#include <iostream>
+#include <set>
+#include <vector>
+
+constexpr int NMAX = 200005;
+int rez[NMAX];
+std::vector<std::vector <int>> G(NMAX);
+//tinem minte cate un set pentru fiecare nod
+std::vector<std::set <int>> s(NMAX);
+
+void dfs(int nod, int t) {
+    for (auto x : G[nod]) {
+        if (x != t) {
+            dfs(x, nod);
+            //daca numarul de elemente din setul fiului este mai mare decat cel din setul nodului, le interschimbam
+            if(s[x].size() > s[nod].size()){
+                s[nod].swap(s[x]);
+            }
+            //parcurgem setul fiecarui fiu si adaugam elementele in setul nodului
+            for (auto x2 : s[x]){
+                s[nod].insert(x2);
+            }
+        }
+    }
+    //obtinem rezultatul folosind metoda .size()
+    rez[nod] = s[nod].size();
+}
+
+int main() {
+    int n, i, u, v, c;
+    std::cin >> n;
+    for (i = 1; i <= n; i++) {
+        std::cin >> c;
+        //valoarea nodului o putem adauga direct in setul acestuia
+        s[i].insert(c);
+    }
+    for (i = 1; i < n; i++) {
+        std::cin >> u >> v;
+        G[u].push_back(v);
+        G[v].push_back(u);
+    }
+    dfs(1, 0);
+    for (i = 1; i <= n; i++) {
+        std::cout << rez[i] << " ";
+    }
+    return 0;
+}
+
+```
+
+Această soluție se încadrează în limitele problemei și va lua punctaj maxim.
+
+
+## Small to large folosind `__gnu_pbds`
+
+Această structură de date oferă aceleași funcții ca structura `set` din STL, însă mai oferă încă 2 funcții care ne pot fi de ajutor în anumite probleme:
+
+1. `find_by_order(x)`, care returnează un iterator către a x-a cheie ca ordine, în timp $O(\log(n))$.
+2. `order_of_key(x)`, care returnează numărul de elemente **strict** mai mici ca x, tot în timp $O(\log(n))$.
+
+Pentru a putea folosi această structură, va trebui să adăugăm la programul nostru următoarele linii:
+```c++
+#include <ext/pb_ds/assoc_container.hpp>
+#include <ext/pb_ds/tree_policy.hpp>
+using namespace __gnu_pbds;
+using ordered_set = tree<int, null_type, std::less_equal<int>, rb_tree_tag, tree_order_statistics_node_update>; 
+```
+În interiorul parantezelor ascuțite de după `tree`, primul câmp reprezintă tipul de date, în acest caz `int`, al patrulea reprezintă tipul de arbore folosit, în acest caz, `rb_tree`, care garantează complexitatea logaritmică a operațiilor de inserare și ștergere. Al treilea câmp reprezintă tipul de comparator folosit, unde `std::less_equal` permite existența a mai multor elemente egale (ca un `multiset`), iar `std::less` nu (ca un `set`).
```

Comment:
Consider ca merge undeva sa pui o observatie ca in ciuda avantajelor pe care le ofera PBDS, constanta este foarte mare si trebuie avut grija cand e folosit in alte contexte (eventual cand @overanalytcl sau altcineva face articol despre STL, trebuie sa punem si pbds ca sa stie lumea sa nu-l foloseasca ca si cum ar "da cu ciocanul" pentru a evita 2-3 minute in plus de gandit la unele probleme mai usoare). Nu priveste articolul de aici dar la normalizari si queryuri de genul asta nu e mereu cea mai buna solutie.

Reply:
> Consider ca merge undeva sa pui o observatie ca in ciuda avantajelor pe care le ofera PBDS, constanta este foarte mare si trebuie avut grija cand e folosit in alte contexte (eventual cand @overanalytcl sau altcineva face articol despre STL, trebuie sa punem si pbds ca sa stie lumea sa nu-l foloseasca ca si cum ar "da cu ciocanul" pentru a evita 2-3 minute in plus de gandit la unele probleme mai usoare). Nu priveste articolul de aici dar la normalizari si queryuri de genul asta nu e mereu cea mai buna solutie.

Problema e că trebuie să menționez caveat-ul că pbds este doar pe GCC (poate și Clang), deci nu îl ai pe MSVC (nu toți folosesc GCC, vezi CF). Spre surprinderea celor care vor vedea pbds, există și chestii pe care compilatoarele nu le au standard (GCC și MSVC sînt cele mai bune exemple). Eu nu aș vrea să includ pbds că eu nu-l consider parte din STL, pentru că nu e standard. E ca și cum aș zice de __int128 la tipuri de date, nu e fair.

Reply:
Atunci sa-l punem la structuri de date? For all intents and purposes se comporta ca o structura de date? Poate cel mai bine pomenim prima oara de el la normalizare, ca asta e normalizarea pentru lenesi (cel putin asta a fost prima aplicatie la care l-am vazut folosit)

Reply:
Da, are mai mult sens acolo, deși devine cam vag articolul de care zici. La normalizare cred că își are rost mențiunea și aia e, o zicem ca fapt divers.

Reply:
Aici am zis despre constanta mare si am adaugat si ceva despre interschimbare, dar am zis sa nu adaug si aceasta parte:

> Problema e că trebuie să menționez caveat-ul că pbds este doar pe GCC (poate și Clang), deci nu îl ai pe MSVC (nu toți folosesc GCC, vezi CF). Spre surprinderea celor care vor vedea pbds, există și chestii pe care compilatoarele nu le au standard (GCC și MSVC sînt cele mai bune exemple). Eu nu aș vrea să includ pbds că eu nu-l consider parte din STL, pentru că nu e standard. E ca și cum aș zice de __int128 la tipuri de date, nu e fair.